### ISSUE TITLE
Internal error when using a specific session.flash (all others work fine)

### ISSUE THREAD

**Issue by jcrmatos** [web2py/web2py#2119]
When running the code below in web2py 2.18.2 returns an internal error.
It is only with this session.flash that web2py returns the error. All others work fine.
If I comment it everything works.
I even tried something simple like session.flash = 'Test' but same result.

```
@auth.requires(lambda: (auth.requires_login() and request.env.http_referer
                        and '/single_prod_opt' in request.env.http_referer))
def on_delete(table, rec_id):
    """Used in the grid."""
    opt_id = db.prod_opt[rec_id].opt_id
    rows = db(db.prod_opt.product_id == session.rec_id).select()
    for row in rows:
        if db.opt[row.opt_id].depends_on_opt == opt_id:
            session.flash = T(
                f'Delete option {db.opt[row.opt_id].name} before deleting this one.')
            break
    else:
        row = db.prod_opt[id_]
        row.update_record(cancel_approved_by=session.auth.user.id,
                          canceled_by=session.auth.user.id, canceled_on=request.now,
                          is_active=False)

    redirect(URL('index', args=[session.rec_id], user_signature=True))
```

Windows 7 Pro x64 SP1+all updates
Python 3.7.1 x86
web2py 2.18.2
Firefox 65.0.2 x64

Here is the error
```
Traceback (most recent call last):
  File "C:\web2py\gluon\main.py", line 444, in wsgibase
    serve_controller(request, response, session)
  File "C:\web2py\gluon\main.py", line 206, in serve_controller
    raise HTTP(response.status, page, **response.headers)
gluon.http.HTTP: 200 OK

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\web2py\gluon\main.py", line 488, in wsgibase
    urllib_quote(xmlescape(response.flash).replace(b'\n', b''))
NameError: name 'urllib_quote' is not defined
```

Thanks,

JM

**Comment by Kkeller83** [comment]
I can not reproduce this with the "Test" message. Maybe something in the translate in combination with flash messages?

**Comment by jcrmatos** [comment]
I tried with a simple session.flash = 'Test' (thus without the T) and it shows the same internal error.

**Comment by jcrmatos** [comment]
Even after the kludge I mentioned to "solve" issue #2110, it still shows the internal error.

**Comment by jcrmatos** [comment]
The solution is simple (or so it appears). Just change the line
`from gluon._compat import Cookie, urllib2`
to
`from gluon._compat import Cookie, urllib2, urllib_quote`
in main.py.

Can anyone create a PR for this?

**Comment by Kkeller83** [comment]
![error4](https://user-images.githubusercontent.com/7712288/53700304-35c49880-3df1-11e9-92f5-b16e8b487420.png)
![error5](https://user-images.githubusercontent.com/7712288/53700305-35c49880-3df1-11e9-9655-be4e9738fdb2.png)

After comparing the files it looks like that the new functional xmlescape that is replacing local_html_escape is not doing its job it's supposed to

**Comment by Kkeller83** [comment]
the error must be somewhere in the yatl/sanitizer.py and in the xmlescape function

**Comment by jcrmatos** [comment]
The yatl/sanitizer.py is the same between both versions.

**Comment by Kkeller83** [comment]
That may be, but in 2.18.1 the local_html_escape function is still used while in 2.18.2 we make the switch to to actually using xmlescape from yatl instead of local_html_escape.

**Comment by Kkeller83** [comment]
if PY2:
    from cgi import escape
    from HTMLParser import HTMLParser
    import urlparse
    from htmlentitydefs import entitydefs, name2codepoint
    basestring = basestring    
else:
    from html import escape
    from html.parser import HTMLParser
    from urllib import parse as urlparse
    from html.entities import entitydefs, name2codepoint
    basestring = str

def xmlescape(text, quote=True, colon=True):
    if not isinstance(text, basestring):
        text = str(text)
    data = escape(text, quote)
    if quote:
        data = data.replace("'", "&#x27;")
        data = data.replace('"', "&quot;")
    if colon:
        data = data.replace(':', '&#58;')
    return data

**Comment by Kkeller83** [comment]
def local_html_escape(data, quote=False):
    """
    Works with bytes.
    Replace special characters "&", "<" and ">" to HTML-safe sequences.
    If the optional flag quote is true (the default), the quotation mark
    characters, both double quote (") and single quote (') characters are also
    translated.
    """
    if PY2:
        import cgi
        data = cgi.escape(data, quote)
        return data.replace("'", "&#x27;") if quote else data
    else:
        import html
        if isinstance(data, str):
            return html.escape(data, quote=quote)
        data = data.replace(b"&", b"&amp;")  # Must be done first!
        data = data.replace(b"<", b"&lt;")
        data = data.replace(b">", b"&gt;")
        data = data.replace(b">", b"&gt;")
        if quote:
            data = data.replace(b'"', b"&quot;")
            data = data.replace(b'\'', b"&#x27;")
        return data

**Comment by Kkeller83** [comment]
comparing the 2 it seems that xmlescape is not taking care of the "b". Maybe it was only tested with Python2?

**Comment by jcrmatos** [comment]
Yes, I saw that, but because I have no idea what all of that does and the consequences of any change, I requested (on the title of this issue) that someone with higher knowledge do the PR.

**Comment by Kkeller83** [comment]
Yeah, fair enough. I fiddled around with it now, but also dont know what I am doing and cant get it to work.
 There seems to be more than just xmlescape fixing required.
The new TranslatorFactory function also seems to have do something with it. 

I will use your "kludge" solution for now meanwhile.
Thanks for figuring this out!

This error must have been introduced with the move to YATL

**Comment by jcrmatos** [comment]
Solved in 2.18.3
